### PR TITLE
Add support for selinux types in manifest files

### PR DESCRIPTION
**Pull request by inorick**
Sextant now supports adding selinux contexts to container files during packing.

An example entry in a manifest file looks like this:

```
selinux:
  type: northstar_t
```

**TODO / open points:**
 - Retag existing files or create copies?
 - Test on hardware

### CODE REVIEW

File: northstar/src/npk/npk.rs [hunk 1]
```diff
@@ -748,6 +758,92 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
...
+fn set_selinux_contexts(root: &Path, context_type: &NonNullString) -> Result<(), Error> {
+    for dir_entry in WalkDir::new(root)
+        .follow_links(false)
+        .into_iter()
+        .filter_map(Result::ok)
+    {
+        let path = dir_entry.path();
+        if let Ok(path_c) = std::ffi::CString::new(path.as_os_str().as_bytes()) {
+            // Read existing selinux context
+            let mut context_buf = [0u8; 256];
```
Comment (flxo) [inline northstar/src/npk/npk.rs]: Why 256? Is this the max length of xattr values?

File: northstar/src/npk/npk.rs [hunk 2]
```diff
@@ -748,6 +758,92 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
     )
 }
 
+fn set_selinux_contexts(root: &Path, context_type: &NonNullString) -> Result<(), Error> {
```
Comment (flxo) [inline northstar/src/npk/npk.rs:761]: Short description what is done in here would be great.

File: northstar/src/npk/manifest.rs [hunk 3]
```diff
@@ -47,10 +47,12 @@ pub struct Manifest {
     pub env: Option<HashMap<NonNullString, NonNullString>>,
     /// Autostart this container upon northstar startup
     pub autostart: Option<Autostart>,
-    /// CGroup config
+    /// CGroup configuration
     pub cgroups: Option<cgroups::CGroups>,
     /// Seccomp configuration
     pub seccomp: Option<Seccomp>,
+    /// SELinux configuration
+    pub selinux: Option<Selinux>,
```
Comment (flxo) [inline northstar/src/npk/manifest.rs:55]: Can we add some checks in `Manifest::verify`? There are for sure restrictions on the label

File: northstar/src/npk/npk.rs [hunk 4]
```diff
@@ -748,6 +758,92 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
...
+                    "Failed to convert old selinux context to C string: {}",
+                    &path.display()
+                )));
+            }
+        } else {
+            return Err(Error::Selinux(format!(
+                "Failed to convert path to C string: {}",
+                &path.display()
+            )));
+        }
```
Comment (flxo) [inline northstar/src/npk/npk.rs]: Is it possible to flatten this by creating some bindings. This is really hard to read and the errors are very similar. Maybe move it to a fn and just use a single error message.

File: northstar/src/seccomp/types.rs [hunk 5]
```diff
@@ -23,6 +23,14 @@ pub struct Seccomp {
     pub allow: Option<HashMap<NonNullString, SyscallRule>>,
 }
 
+/// SELinux configuration
+#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct Selinux {
+    /// Explicit list of allowed syscalls
+    #[serde(rename = "type")]
+    pub context_type: NonNullString,
```
Comment (flxo) [inline northstar/src/seccomp/types.rs:31]: Can we remove the `rename`? Either use `context_type` in them yaml or `r#type` as field name. I don't know what's the better choice.

File: northstar/src/npk/npk.rs (hunk 1 above)
Comment (inorick) [reply northstar/src/npk/npk.rs]: We now use another call to getxattr to determine the proper size.

File: northstar/src/npk/npk.rs (hunk 4 above)
Comment (inorick) [reply northstar/src/npk/npk.rs]: I broke the steps down into sub-functions that only return one error each.

File: northstar/src/seccomp/types.rs (hunk 5 above)
Comment (inorick) [reply northstar/src/seccomp/types.rs:30]: context_type it is.

File: northstar/src/npk/manifest.rs [hunk 6]
```diff
@@ -150,6 +152,20 @@ impl Manifest {
                 _ => Ok(()),
             })?;
 
+        // Check selinux context type
+        if let Some(selinux) = &self.selinux {
```
Comment (flxo) [inline northstar/src/npk/manifest.rs:156]: Is there a length restriction on the labels?

File: northstar/src/npk/npk.rs [hunk 7]
```diff
@@ -748,6 +758,107 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
...
+    if components.len() >= 3 {
+        components[2] = context_type;
+        CString::new(components.join(":").as_bytes()).ok()
+    } else {
+        None
+    }
+}
+
+/// Write selinux context to a path
+fn write_selinux_context(path_c: &CString, context: &CString) -> bool {
```
Comment (flxo) [inline northstar/src/npk/npk.rs:849]: Why is a `CString` used here? I would have expected a `&std::path::Path` or at least a `&CStr`.
`&CString` doesn't make sense.

File: northstar/src/npk/npk.rs [hunk 8]
```diff
@@ -748,6 +758,107 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
...
+    let name_c = CString::new("security.selinux").unwrap();
```
Comment (flxo) [inline northstar/src/npk/npk.rs:850]: This can be a `const`

File: northstar/src/npk/npk.rs [hunk 9]
```diff
@@ -748,6 +758,107 @@ fn sign_hashes(key_pair: &Keypair, hashes_yaml: &str) -> String {
...
+    let context = CStr::from_bytes_with_nul(&context_buf[..size.try_into().ok()?])
+        .ok()?
+        .to_str()
+        .ok()?
+        .to_string();
+    Some(context)
+}
+
+/// Replace the tag of a selinux context and return the updated version
+fn replace_selinux_context_type(context: &str, context_type: &NonNullString) -> Option<CString> {
```
Comment (flxo) [inline northstar/src/npk/npk.rs:838]: I more and more doubt that `NonNullString` was a good idea or not...

File: northstar/src/npk/npk.rs (hunk 7 above)
Comment (flxo) [inline northstar/src/npk/npk.rs:849]: Return a `Result` instead of bool here.

File: northstar/src/npk/manifest.rs (hunk 6 above)
Comment (inorick) [reply northstar/src/npk/manifest.rs:156]: I think we can use
```
#define XATTR_SIZE_MAX 65536
```
(https://www.spinics.net/lists/selinux/msg09373.html)

File: northstar/src/npk/npk.rs (hunk 8 above)
Comment (inorick) [reply northstar/src/npk/npk.rs:850]: CString::new is not a const function :/

Comment (flxo) [reply northstar/src/npk/npk.rs]: Right. But we still could improve it to use https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul_unchecked

this avoids a copy.

**Pull request merged**